### DISCUSSION TITLE
Touch Pin as Button - Max count - ESP32-S2 / S3

### DISCUSSION THREAD

**Discussion by nccchirag** [arendst/Tasmota#16518] [Q&A] [answered]
ESP32-S2 / S3 have 14 capacitive touch sensors/pins. How can I use all of them with tasmota?
As per tasmota docs & source the max button limit is capped at 8 (MAX_KEYS) 
**Button_tc1...8	Touch pin as button**
const uint8_t MAX_KEYS = 8;                 // Max number of keys or buttons (up to 28)

Can I increase the same or I need to modify source for achieving the support?

#if defined(CONFIG_IDF_TARGET_ESP32S2) || defined(CONFIG_IDF_TARGET_ESP32S3)
    const uint8_t MAX_KEYS = 14;
#endif

Please guide the right way to implement it

Thank you

**Comment by arendst** [comment, marked as answer]
I think if you change ``tasmota.h`` lines 48+ to the below it should/could work.

```
#ifdef ESP8266
const uint8_t MAX_RELAYS = 8;               // Max number of relays (up to 28)
const uint8_t MAX_INTERLOCKS = 4;           // Max number of interlock groups (up to MAX_INTERLOCKS_SET)
const uint8_t MAX_SWITCHES = 8;             // Max number of switches (up to MAX_SWITCHES_SET)
const uint8_t MAX_KEYS = 8;                 // Max number of keys or buttons (up to 28)
#endif  // ESP8266
#ifdef ESP32
const uint8_t MAX_RELAYS = 28;              // Max number of relays (up to 28)
const uint8_t MAX_INTERLOCKS = 14;          // Max number of interlock groups (up to MAX_INTERLOCKS_SET)
const uint8_t MAX_SWITCHES = 28;            // Max number of switches (up to MAX_SWITCHES_SET)
const uint8_t MAX_KEYS = 28;                // Max number of keys or buttons (up to 28)
#endif  // ESP32
```

My first investigation shows it might work as all bitmasks look like 32-bits. Give it a try

**Reply by nccchirag** [reply]
thank you so much for the prompt response!
sure, let me check

I would like to take this opportunity to thank you @arendst for creating this wonderful software & also to all the maintainers of tasmota 🙏
your work is commendable and inspiring✨

**Reply by nccchirag** [reply]
Do I need to define in the user_config_override.h for touch to be supported on ESP32S3?

#undef BUTTON_V3
#define BUTTON_V3

Also, which one of the two (_BUTTON_V2 vs V3_) are better for touch application?

Thank You

**Reply by arendst** [reply]
No.

Do not change anything with regards to BUTTON_V3 or _V2.

V3 is the latest and is supposed to be the future proof one. Touch support has not changed between the two versions.

**Reply by nccchirag** [reply]
I'm able to compile the latest dev build using tasmota32s3 env but after configuring the module with Button_tc, none of the touch sensors seem to work! 

am I missing any settings / configuration, please guide

Thank you

**Dev Environment**
Ai-Thinker ESP32 Dev Kit - **ESP-S3-12K, 8M flash & PSRAM**
Windows x64 PC
VsCode & PlatformIO latest
latest dev branch of tasmota
GPIO1-14 are touch sensors in ESP32S3 

![image](https://user-images.githubusercontent.com/7569367/190348370-d8fe052b-4740-43f2-9742-a3e64e7232d4.png)

![image](https://user-images.githubusercontent.com/7569367/190348699-389d0ecf-24cf-49a3-8e19-015e32e9e960.png)

**Template**
{"NAME":"ESP32S3","GPIO":[0,5408,5409,5410,5411,5412,5413,5414,5415,5416,5417,5418,5419,5420,45,0,0,0,0,0,0,0,0,0,0,0,0,0,576,0,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":1}

![image](https://user-images.githubusercontent.com/7569367/190348907-7bffd0bf-c399-4a45-8c9c-061e59e5843e.png)

**Reply by arendst** [reply]
Sorry, can't help with touch inputs. Never used them. Someone else maybe...

**Reply by nccchirag** [reply]
The issue seems to be with touch threshold which is defined to be < 256 (_line 2623 support_command.ino_)

10:07:50.908 _value < TOUCH_BUTTON.pin_threshold: 3 23369 255
10:07:50.910 _value < TOUCH_BUTTON.pin_threshold: 4 15959 255

commenting the if condition allows me to set threshold to the desired value, also the struct needs to be modified

![image](https://user-images.githubusercontent.com/7569367/190365982-750f4bfa-09d7-49c8-bfb5-38043f3affe6.png)

Also, the serialplotter is designed to filter touch buttons numbered from 1..4. It needs to be modified to support Touch points 1..14

Thank you

**Reply by nccchirag** [reply]
> Sorry, can't help with touch inputs. Never used them. Someone else maybe...

No problem thank you

**Reply by arendst** [reply]
Except the plotter the issues should be fixed in latest dev just now.

**Comment by arendst** [comment]
Implemented just now in latest dev release.

**Reply by nccchirag** [reply]
this was my second change request and tasmota team has provided an immediate solution/response as always 🙏
sincere thanks to @arendst @s-hadinger @barbudor & team tasmota!

**Comment by SimonFili** [comment]
Hi, @nccchirag,  would you have any insight on this Q&A ?
https://github.com/arendst/Tasmota/discussions/16596

**Reply by nccchirag** [reply]
https://github.com/arendst/Tasmota/discussions/16596#discussioncomment-3700029
